routingpy/routers/heremaps.py

```diff
@@ -88,14 +90,19 @@ def __init__(
         :type skip_api_error: bool
         """
 
-        if app_id is None and app_code is None:
-            raise KeyError("HERE Maps app_id and app_code must be specified.")
+        if app_id is None and app_code is None and api_key is None:
+            raise KeyError("HERE Maps app_id and app_code, or api_key must be specified.")
+        if (app_id is not None or app_code is not None) and api_key is not None:
+            raise KeyError("Either HERE Maps app_id and app_code, or api_key can be specified, not both.")
+
 
         self.app_code = app_code
         self.app_id = app_id
+        self.api_key = api_key
```

Comment:
Let's avoid all the conditional calls around apikey by smth like:

```
if self.api_key:
    self.base_url = self._APIKEY_BASE_URL
    self.auth = {"apikey": self.api_key}
else:
    self.base_url = self._DEFAULT_BASE_URL
    self.auth = {"app_id": self.app_id, "app_code": self.app_code}
```

Then in the methods you could just reference e.g. `params = deepcopy(self.auth)` to initialize the parameters and `self.base_url` in the `__init__`. 

Would make it a lot cleaner IMHO. What do you think?

Reply:
Hi, yep absolutely agree, just wanted to keep my changes to the codebase minimal given that idk what the whole situation with switching over to heremaps 8.x is. 

Does this the params assignment need to be a deepcopy? I think a shallow copy should be fine for this case no?

Along this same line of thinking, should the different base urls for each request type (routing, isochrone, matrix ect) all be set in init as well? otherwise i don't see a way around the checks on line 590, 990, ect.